src/expire.c

```diff
@@ -567,19 +571,20 @@ void pexpireatCommand(client *c) {
 }
 
 /* Implements TTL and PTTL */
-void ttlGenericCommand(client *c, int output_ms) {
+void ttlGenericCommand(client *c, int output_ms, int output_abs) {
     long long expire, ttl = -1;
 
     /* If the key does not exist at all, return -2 */
     if (lookupKeyReadWithFlags(c->db,c->argv[1],LOOKUP_NOTOUCH) == NULL) {
         addReplyLongLong(c,-2);
         return;
     }
+
     /* The key exists. Return -1 if it has no expire, or the actual
      * TTL value otherwise. */
     expire = getExpire(c->db,c->argv[1]);
     if (expire != -1) {
-        ttl = expire-mstime();
+        ttl = output_abs ? expire : expire-mstime();
```

Comment:
I'm not entirely understanding the problem here.

My assumption is that Redis never stores a negative absolute timestamp into the expire dictionary. If `EXPIREAT/PEXPIREAT` is executed with a negative Unix timestamp, it will [result in the key being deleted](https://github.com/redis/redis/blob/92bde124caaedef0e1572765854b2deb9aa22a35/src/expire.c#L524), therefore not stored into the expire dictionary.

And since `EXPIRETIME` just simply returns what is stored in the expire dictionary, there is no legitimate case where it would return a negative value. Therefore it's safe to reserve `-1` and `-2` for error responses.

What am I missing?

Reply:
i guess it's just me.. being obsessed with the fact that negative unix times are valid values.
in redis, we don't store arbitrary dates in the past, just the future, so a valid negative timestamp would only happen if the machine's wall-clock is set to before 1/1/1970, in which case the key won't deleted on the spot by `checkAlreadyExpired`.
i guess this is not a valid use case for redis.

Till now in the API (commands) we mainly abused the values -1 and -2 for relative times, so that's ok either way (even if your clock is before 1970).
We do have that issue in `getExpire` too, work works with absolute times, but this is internal and can in theory be fixed one day.

So we need to realize that the semantics of this new command carves that in stone (API).
@yossigo @madolson WDYT?

Reply:
I'm okay with blocking negative expires in Redis, I can't see a world where we would want to store them and return them.

Reply:
+1 for disallowing setting negative Unix timestamps with `EXPIRE/EXPIREAT/PEXPIRE/PEXPIREAT` altogether.

Reply:
I don't think we wanna change the behavior of existing commands in that case (return an error), but as we concluded above, we can assume negative values will never find their way into the expires dict. 
So I think the current code in this PR is fine in that regard.